File: python/open_guardrail/guards/student_record_guard.py
"""FERPA: prevent exposure of education records."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_STUDENT_RECORD = re.compile(
    r"\b(?:GPA|grade\s+point\s+average"
    r"|transcript|student\s+ID"
    r"|enrollment\s+status"
    r"|disciplinary\s+record"
    r"|financial\s+aid|student\s+loan"
    r"|academic\s+record|test\s+scores"
    r"|\w+'s\s+grades)\b",
    re.IGNORECASE,
)
_SHARE = re.compile(
    r"\b(?:share\s+with|send\s+to"
    r"|post|publish"
    r"|display\s+publicly"
    r"|email\s+to)\b",
    re.IGNORECASE,
)


class _StudentRecordGuard:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "student-record-guard"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        has_record = bool(
            _STUDENT_RECORD.search(text)
        )
        has_share = bool(_SHARE.search(text))
        triggered = has_record and has_share
        elapsed = (
            time.perf_counter() - start
        ) * 1000
        if not triggered:
            return GuardResult(
                guard_name="student-record-guard",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )
        return GuardResult(
            guard_name="student-record-guard",
            passed=False,
            action=self.action,
            message=(
                "Student record sharing detected"
                " — potential FERPA violation"
            ),
            latency_ms=round(elapsed, 2),
            details={
                "student_record": has_record,
                "sharing_pattern": has_share,
                "reason": (
                    "FERPA prohibits unauthorized"
                    " disclosure of student records"
                ),
            },
        )


def student_record_guard(
    *, action: str = "block"
) -> _StudentRecordGuard:
    return _StudentRecordGuard(action=action)

File: packages/guards/src/student-record-guard.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';
interface StudentRecordGuardOptions { action: 'block' | 'warn'; }
const STUDENT_RECORD = /\b(?:GPA|grade\s+point\s+average|transcript|student\s+ID|enrollment\s+status|disciplinary\s+record|financial\s+aid|student\s+loan|academic\s+record|test\s+scores|\w+'s\s+grades)\b/i;
const SHARE = /\b(?:share\s+with|send\s+to|post|publish|display\s+publicly|email\s+to)\b/i;
export function studentRecordGuard(options: StudentRecordGuardOptions): Guard {
  return { name: 'student-record-guard', version: '0.1.0', description: 'FERPA: prevent exposure of education records', category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const hasRecord = STUDENT_RECORD.test(text);
      const hasShare = SHARE.test(text);
      const triggered = hasRecord && hasShare;
      return { guardName: 'student-record-guard', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? 'Student record sharing detected — potential FERPA violation' : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { studentRecordDetected: hasRecord, sharingPatternDetected: hasShare, reason: 'FERPA prohibits unauthorized disclosure of student records' } : undefined,
      };
    },
  };
}
